base/cvd/cuttlefish/host/libs/web/cas/cas_downloader.cpp

```diff
@@ -0,0 +1,298 @@
+                   cas_identifier.cas_instance);
+  cmd.AddParameter("-", std::string(kFlagCasAddr), "=" +
+                   cas_identifier.cas_addr);
+  cmd.AddParameter("-", std::string(kFlagDigest), "=" +
+                   cas_identifier.digest);
+  cmd.AddParameter("-", std::string(kFlagDir), "=", target_directory);
+  if (stats_filepath.has_value()) {
+    cmd.AddParameter("-", std::string(kFlagDumpJson), "=", *stats_filepath);
+  }
+  for (const std::string& flag : flags) {
+    cmd.AddParameter(flag);
+  }
+  return cmd;
+}
+
+}  // namespace
+
+Result<std::unique_ptr<CasDownloader>> CasDownloader::Create(
+    const CasDownloaderFlags& cas_downloader_flags,
+    const std::string& service_account_filepath) {
+  std::string downloader_path = cas_downloader_flags.downloader_path;
+  bool prefer_uncompressed = cas_downloader_flags.prefer_uncompressed;
+  std::vector<std::string> cas_flags;
+
+  std::string config_filepath = cas_downloader_flags.cas_config_filepath;
+  if (config_filepath.empty()) {
+    Json::Value config_flags = ConvertToConfigFlags(cas_downloader_flags);
+    cas_flags = CreateCasFlags(downloader_path, config_flags);
+  } else {
+    std::string config_contents = CF_EXPECT(ReadFileContents(config_filepath));
+    Json::Value config = CF_EXPECT(ParseJson(config_contents));
+    downloader_path = config[kKeyDownloaderPath].asString();
+    prefer_uncompressed = config["prefer-uncompressed"].asBool();
+    cas_flags = CreateCasFlags(downloader_path, config[kKeyFlags]);
+  }
+
+  CF_EXPECTF(FileExists(downloader_path),
+             "CAS Downloader binary not found: '{}'", downloader_path);
+  if (!service_account_filepath.empty() &&
+      FileExists(service_account_filepath)) {
+    cas_flags.push_back("-" + std::string(kFlagServiceAccountJson) + "=" +
+                        std::string(service_account_filepath));
+  } else {
+    cas_flags.push_back("-" + std::string(kFlagUseAdc));
+  }
+
+  return std::unique_ptr<CasDownloader>(
+      new CasDownloader{downloader_path, cas_flags, prefer_uncompressed});
+}
+
+CasDownloader::CasDownloader(std::string downloader_path,
+                             std::vector<std::string> flags,
+                             bool prefer_uncompressed)
+    : downloader_path_(std::move(downloader_path)),
+      flags_(std::move(flags)),
+      prefer_uncompressed_(prefer_uncompressed) {}
+
+Result<void> CasDownloader::DownloadFile(
+    const std::string& build_id, const std::string& build_target,
+    const std::string& artifact_name, const std::string& target_directory,
+    const DigestsFetcher& digests_fetcher,
+    const std::optional<std::string>& stats_filepath) {
+  std::string download_directory = target_directory;
+  CasIdentifier cas_identifier = CF_EXPECT(
+      GetCasIdentifier(build_id, build_target, artifact_name, digests_fetcher));
+  std::string filename = cas_identifier.filename;
+  if (filename.find("_chunked_dir_") == 0) {
```

Comment:
Confirming that this is intentionally checking the beginning of the string, since `0` is the first character and comparing against `std::string::npos` would be checking for presence.

Reply:
Yes. It is a prefix. Added a comment to make it explicit.